diff --git a/gi/overrides/Gtk.py b/gi/overrides/Gtk.py
index 6ddc12f607a596295997b9ecdfb453889c7e29e1..5511d002cbb8e0b4f0f41fe6bb5b7950f20bb136 100644
--- a/gi/overrides/Gtk.py
+++ b/gi/overrides/Gtk.py
@@ -1643,6 +1643,26 @@ class TreeModelFilter(Gtk.TreeModelFilter):
 TreeModelFilter = override(TreeModelFilter)
 __all__.append('TreeModelFilter')
 
+
+class CssProvider(Gtk.CssProvider):
+    def load_from_data(self, text, length=-1):
+        if (Gtk.get_major_version(), Gtk.get_minor_version()) >= (4, 9):
+            if isinstance(text, bytes):
+                text = text.decode("utf-8")
+            if not isinstance(text, str):
+                raise TypeError('text must be a string')
+            super(CssProvider, self).load_from_data(text, length)
+        else:
+            if isinstance(text, str):
+                text = text.encode("utf-8")
+            if not isinstance(text, bytes):
+                raise TypeError('text must be bytes')
+            super(CssProvider, self).load_from_data(text)
+
+
+CssProvider = override(CssProvider)
+__all__.append("CssProvider")
+
 if GTK4:
     class CustomSorter(Gtk.CustomSorter):
 
diff --git a/tests/test_overrides_gtk.py b/tests/test_overrides_gtk.py
index 50c1ad8bdb5a171f877aef39f2db1bcc386f01fa..1b3e93ba1f99288d60b3c895370adafd203cd4d8 100644
--- a/tests/test_overrides_gtk.py
+++ b/tests/test_overrides_gtk.py
@@ -3026,3 +3026,15 @@ def test_button_focus_on_click():
     assert b.get_focus_on_click()
     b.set_focus_on_click(False)
     assert not b.get_focus_on_click()
+
+
+@pytest.mark.parametrize(
+    "data",
+    [
+        "* { background: white; }",
+        "* { background: white; }".encode()
+    ]
+)
+def test_css_provider_load_from_data(data):
+    provider = Gtk.CssProvider.new()
+    provider.load_from_data(data)
